src/Monotonic.cpp

```diff
@@ -67,135 +214,109 @@ class MonotonicVisitor : public IRVisitor {
 
         // A narrowing cast. There may be more cases we can catch, but
         // for now we punt.
-        if (result != Monotonic::Constant) {
-            result = Monotonic::Unknown;
+        if (!is_constant(result)) {
+            result = ConstantInterval::everything();
         }
     }
 
     void visit(const Variable *op) override {
         if (op->name == var) {
-            result = Monotonic::Increasing;
+            result = ConstantInterval::single_point(1);
         } else if (scope.contains(op->name)) {
             result = scope.get(op->name);
         } else {
-            result = Monotonic::Constant;
+            result = ConstantInterval::single_point(0);
         }
     }
 
-    Monotonic flip(Monotonic r) {
-        switch (r) {
-        case Monotonic::Increasing:
-            return Monotonic::Decreasing;
-        case Monotonic::Decreasing:
-            return Monotonic::Increasing;
-        default:
-            return r;
-        }
-    }
-
-    Monotonic unify(Monotonic a, Monotonic b) {
-        if (a == b) {
-            return a;
-        }
-
-        if (a == Monotonic::Unknown || b == Monotonic::Unknown) {
-            return Monotonic::Unknown;
-        }
-
-        if (a == Monotonic::Constant) {
-            return b;
-        }
-
-        if (b == Monotonic::Constant) {
-            return a;
-        }
-
-        return Monotonic::Unknown;
-    }
-
     void visit(const Add *op) override {
         op->a.accept(this);
-        Monotonic ra = result;
+        ConstantInterval ra = result;
         op->b.accept(this);
-        Monotonic rb = result;
-        result = unify(ra, rb);
+        ConstantInterval rb = result;
+        result = add(ra, rb);
     }
 
     void visit(const Sub *op) override {
         op->a.accept(this);
-        Monotonic ra = result;
+        ConstantInterval ra = result;
         op->b.accept(this);
-        Monotonic rb = result;
-        result = unify(ra, flip(rb));
+        ConstantInterval rb = result;
+        result = sub(ra, rb);
     }
 
     void visit(const Mul *op) override {
-        op->a.accept(this);
-        Monotonic ra = result;
-        op->b.accept(this);
-        Monotonic rb = result;
-
-        if (ra == Monotonic::Constant && rb == Monotonic::Constant) {
-            result = Monotonic::Constant;
-        } else if (is_positive_const(op->a)) {
-            result = rb;
-        } else if (is_positive_const(op->b)) {
-            result = ra;
-        } else if (is_negative_const(op->a)) {
-            result = flip(rb);
-        } else if (is_negative_const(op->b)) {
-            result = flip(ra);
+        if (op->type.is_scalar()) {
+            op->a.accept(this);
+            ConstantInterval ra = result;
+            op->b.accept(this);
+            ConstantInterval rb = result;
+
+            if (const int64_t *b = as_const_int(op->b)) {
```

Comment:
Why isn't `ConstantInterval multiply(const ConstantInterval &a, const ConstantInterval &b)` used here?

Reply:
b is an Expr, not a ConstantInterval, and DerivativeBounds loses information. I think correct thing to do here is something like the product rule for derivatives:

    a * DerivativeBounds(b) + b * DerivativeBounds(a)

But now that the intervals are ConstantIntervals, that's kind of a pain to implement, and I believe it reduces to the same thing as the logic here (when a or b are constants).